### PR TITLE
DM

### PR DESCRIPTION
**Pull request by Ig-dolci**

### CODE REVIEW

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 1]
```diff
@@ -733,7 +1765,16 @@ cdef class DM(Object):
...
 
-    def addCoarsenHook(self, coarsenhook, restricthook, args=None, kargs=None):
+    def addCoarsenHook(self, coarsenhook, restricthook, args=None, kargs=None) -> None:
+        """Add a callback to be executed when restricting a nonlinear problem to a coarse grid.
+        
+        Logically Collective; No Fortran Support.
+
+        See Also
+        --------
+        petsc.DMCoarsenHookAdd
```
Comment (dham) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1775]: missing parameters documentation.

**Review by dham — CHANGES_REQUESTED** [review]

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 2]
```diff
@@ -671,14 +1871,49 @@ cdef class DM(Object):
     createLocalVector = createLocalVec
     getMatrix = createMatrix = createMat
 
-    def setKSPComputeOperators(self, operators, args=None, kargs=None):
```
Comment (Ig-dolci) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:674]: operators type will be defined in KSP PR.

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 3]
```diff
@@ -495,7 +1278,14 @@ cdef class DM(Object):
         CHKERR( DMAdaptLabel(self.dm, clbl, &newdm.dm) )
         return newdm
 
-    def adaptMetric(self, Vec metric, bdLabel=None, rgLabel=None):
+    def adaptMetric(self, Vec metric, bdLabel: str | None, rgLabel: str | None) -> DM:
+        """Return a mesh adapted to the specified metric field.
+
+        See Also
+        --------
+        petsc.DMAdaptMetric
```
[the remark this answers is not in the document]
Comment (Ig-dolci) [reply src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1390]: Done.

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 4]
```diff
@@ -542,124 +1369,298 @@ cdef class DM(Object):
...
 
-    def clearLabelValue(self, name, point, value):
+    def clearLabelValue(self, name: str, point: int, value: int) -> None:
+        """Remove a point from a `DMLabel` with given value.
+
+        Not Collective.
+
+        See Also
+        --------
+        petsc.DMClearLabelValue
```
[the remark this answers is not in the document]
Comment (Ig-dolci) [reply src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1706]: Done.

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 5]
```diff
@@ -184,7 +490,27 @@ cdef class DM(Object):
...
+        """Set an auxiliary vector for a specific region.
+
+        Not Collective.
+
+        Parameters
+        ----------
+        aux
+            The auxiliary vector.
+        label
+            The `DM` label.
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:503]:
```suggestion
            The name of the `DMLabel`.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 6]
```diff
@@ -184,7 +490,27 @@ cdef class DM(Object):
...
+        value : int, optional
+            Indicate the region.
+        part : int, optional
+            The equation part, or 0 is unused.
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:507]:
```suggestion
        value
            Indicate the region.
        part
            The equation part, or 0 is unused.
```

**Review by ksagiyam — CHANGES_REQUESTED** [review]

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 7]
```diff
@@ -194,7 +524,25 @@ cdef class DM(Object):
...
+
+        See Also
+        --------
+        petsc.DMGetLabel, petsc.DMGetAuxiliaryVec
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:545]:
```suggestion
    def getAuxiliaryVec(self, label: str | None = None, value: int | None = 0, part: int | None = 0) -> Vec:
        """Return an auxiliary vector for region.

        Not collective.

        Parameters
        ----------
        label
            The name of the `DMLabel`.
        value
            Indicate the region.
        part
            The equation part, or 0 is unused.

        See Also
        --------
        DM.getLabel, petsc.DMGetAuxiliaryVec

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 8]
```diff
@@ -206,23 +554,71 @@ cdef class DM(Object):
...
+        index
+            The field number.
+        field
+            The discretization object.
+
+        See Also
+        --------
+        petsc.DMSetField
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:600]:
```suggestion
    def setField(self, index: int, Object field, label: str | None = None) -> None:
        """Set the discretization object for a given `DM` field.

        Logically collective.

        Parameters
        ----------
        index
            The field number.
        field
            The discretization object.
        label
            The name of the label indicating the support of the field, or `None` for the entire mesh, currently must be `None`.

        See Also
        --------
        petsc.DMSetField

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 9]
```diff
@@ -206,23 +554,71 @@ cdef class DM(Object):
...
+        Parameters
+        ----------
+        index
+            The field number
+
+        See Also
+        --------
+        petsc.DMGetField
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:621]:
```suggestion
    def getField(self, index: int) -> tuple[Object, None]:
        """Return the discretization object for a given `DM` field.
        
        Not collective.
        
        Parameters
        ----------
        index
            The field number.

        See Also
        --------
        petsc.DMGetField

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 10]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
         cdef Object field = subtype_Object(cobj)()
         field.obj[0] = cobj
         PetscINCREF(field.obj)
-        return (field, None)
+        return (field, None) # TODO REVIEW
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:630]:
```suggestion
        return (field, None)
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 11]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+
+    def addField(self, Object field, label=None) -> None:
+        """Add a field to a `DM` object.
+
+        Logically collective.
+
+        Parameters
+        ---------
+        field
+            The discretization object.
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:640]:
```suggestion
    def addField(self, Object field, label: str | None = None) -> None:
        """Add a field to a `DM` object.

        Logically collective.

        Parameters
        ---------
        field
            The discretization object.
        label
            The name of the label indicating the support of the field, or `None` for the entire mesh, currently must be `None`.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 12]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
         cdef PetscObject  cobj = field.obj[0]
         cdef PetscDMLabel clbl = NULL
         assert label is None
         CHKERR( DMAddField(self.dm, clbl, cobj) )
 
-    def clearFields(self):
+    def clearFields(self) -> None:
+        """Remove all fields from the `DM`.
+
+        Logically Collective
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:657]:
```suggestion
        Logically collective.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 13]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def copyFields(self, DM dm) -> None:
+        """Copy the discretizations of a `DM` into another `DM`.
+
+        Collective.
+
+        See Also
+        --------
+        petsc.DMCopyFields
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:677]:
```suggestion
    def copyFields(self, DM dm) -> None:
        """Copy the discretizations of this `DM` into another `DM`.

        Collective.

        Parameters
        ----------
        dm
            The `DM` that the fields are copied into.

        See Also
        --------
        petsc.DMCopyFields

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 14]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def copyDisc(self, DM dm) -> None:
+        """Copy fields and discrete systems of a `DM` into another `DM`.
+
+        Collective.
+
+        See Also
+        --------
+        petsc.DMCopyDisc
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:750]:
```suggestion
    def copyDisc(self, DM dm) -> None:
        """Copy fields and discrete systems of a `DM` into another `DM`.

        Collective.

        Parameters
        ----------
        dm
            The `DM` that the fields and discrete systems are copied into.

        See Also
        --------
        petsc.DMCopyDisc

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 15]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
         CHKERR( DMCopyDisc(self.dm, dm.dm) )
 
     #
 
-    def getBlockSize(self):
+    def getBlockSize(self) -> int:
+        """Return the inherent block size associated with a `DM`.
+
+        Not Collective
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:758]:
```suggestion
        Not collective.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 16]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def setVecType(self, vec_type: str) -> None:
+        """Set the type of vector.
+
+        Logically collective.
+
+        See Also
+        --------
+        petsc.DMSetVecType
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:783]:
```suggestion
    def setVecType(self, vec_type: Vec.Type | str) -> None:
        """Set the type of vector.

        Logically collective.

        See Also
        --------
        Vec.Type, petsc.DMSetVecType

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 17]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def restoreGlobalVec(self, Vec vg) -> None:
+        """Restore a global vector.
+
+        Not collective.
+
+        See Also
+        --------
+        petsc.DMRestoreGlobalVector
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:840]:
```suggestion
    def restoreGlobalVec(self, Vec vg) -> None:
        """Restore a global vector.

        Not collective.

        Parameters
        ----------
        vg
            The global vector.

        See Also
        --------
        petsc.DMRestoreGlobalVector

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 18]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
         CHKERR( PetscObjectDereference(<PetscObject>vg.vec) )
         CHKERR( DMRestoreGlobalVector(self.dm, &vg.vec) )
 
-    def getLocalVec(self):
+    def getLocalVec(self) -> Vec:
+        """Return a local vector.
+
+        Not collective
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:847]:
```suggestion
        Not collective.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 19]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+        vg
+            The global vector.
+        vl
+            The local vector.
+
+        See Also
+        --------
+        petsc.DMGlobalToLocalBegin, petsc.DMGlobalToLocalEnd
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:898]:
```suggestion
    def globalToLocal(self, Vec vg, Vec vl, addv: InsertMode | None = None) -> None:
        """Update local vectors from global vector.

        Neighborwise collective.

        Parameters
        ----------
        vg
            The global vector.
        vl
            The local vector.
        addv
            `InsertMode.INSERT_VALUES` or `InsertMode.ADD_VALUES`.

        See Also
        --------
        petsc.DMGlobalToLocalBegin, petsc.DMGlobalToLocalEnd

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 20]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+        vg
+            The global vector.
+        vl
+            The local vector.
+
+        See Also
+        --------
+        petsc.DMLocalToLocalBegin, petsc.DMLocalToLocalEnd
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:944]:
```suggestion
    def localToLocal(self, Vec vl, Vec vlg, addv: InsertMode | None = None) -> None:
        """Map the values from a local vector to another local vector.

        Neighborwise collective.

        Parameters
        ----------
        vg
            The global vector.
        vl
            The local vector.
        addv
            `InsertMode.INSERT_VALUES` or `InsertMode.ADD_VALUES`.

        See Also
        --------
        petsc.DMLocalToLocalBegin, petsc.DMLocalToLocalEnd

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 21]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
         cdef Section sec = Section()
         CHKERR( DMGetCoordinateSection(self.dm, &sec.sec) )
         PetscINCREF(sec.obj)
         return sec
 
-    def setCoordinates(self, Vec c):
+    def setCoordinates(self, Vec c) -> None:
+        """Set a global vector that holds the coordinates.
+        Collective
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:998]:
```suggestion
        """Set a global vector that holds the coordinates.

        Collective
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 22]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+
+        Notes
+        -----
+        The option ``-dm_mat_type`` is used to set the matrix type.
+
+        See Also
+        --------
+        petsc_options, petsc.DMSetMatType
+        
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1147]:
```suggestion
    def setMatType(self, mat_type: Mat.Type | str) -> None:
        """Set matrix type to be used by `DM.createMat`.
        
        Logically collective.

        Parameters
        ----------
        mat_type
            The matrix type.

        Notes
        -----
        The option ``-dm_mat_type`` is used to set the matrix type.

        See Also
        --------
        petsc_options, petsc.DMSetMatType
        
        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 23]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def createMassMatrix(self, DM dmf) -> Mat:
+        """Return the mass matrix between two `DM` objects.
+
+        Collective.
+
+        See Also
+        --------
+        petsc.DMCreateMassMatrix
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1175]:
```suggestion
    def createMassMatrix(self, DM dmf) -> Mat:
        """Return the mass matrix between this `DM` and the given `DM`.

        Collective.

        Parameters
        ----------
        dmf
            The second `DM`.

        See Also
        --------
        petsc.DMCreateMassMatrix

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 24]
```diff
@@ -231,191 +627,579 @@ cdef class DM(Object):
...
+    def createInterpolation(self, DM dm) -> tuple[Mat,Vec]:
+        """Return the interpolation matrix between two `DM` objects.
+
+        Collective.
+
+        See Also
+        --------
+        petsc.DMCreateInterpolation
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1194]:
```suggestion
    def createInterpolation(self, DM dm) -> tuple[Mat,Vec]:
        """Return the interpolation matrix between this `DM` and the given `DM`.

        Collective.

        Parameters
        ----------
        dm
           The second, finer `DM`.

        See Also
        --------
        petsc.DMCreateInterpolation

        """
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 25]
```diff
@@ -457,7 +1325,21 @@ cdef class DM(Object):
             hierarchy.append(dmf)
         return hierarchy
 
-    def coarsenHierarchy(self, nlevels):
+    def coarsenHierarchy(self, nlevels: int) -> list:
+        """Return a `DM` object coarsed.
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1329]:
```suggestion
        """Coarsen this `DM` and return the coarsened `DM` hierarchy.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx [hunk 26]
```diff
@@ -470,23 +1352,64 @@ cdef class DM(Object):
...
+    def adaptLabel(self, label: str) -> DM:
+        """Adapt a `DM` based on a `DMLabel`.
+
+        Collective.
+
+        See Also
+        --------
+        petsc.DMAdaptLabel
+
+        """
```
Comment (ksagiyam) [inline src/binding/petsc4py/src/petsc4py/PETSc/DM.pyx:1412]:
```suggestion
    def adaptLabel(self, label: str) -> DM:
        """Adapt a `DM` based on a `DMLabel`.

        Collective.
        
        Parameters
        ----------
        label
            The name of the `DMLabel`.

        See Also
        --------
        petsc.DMAdaptLabel

        """
```

**Review by ksagiyam — DISMISSED** [review]

**Review by dham — APPROVED** [review]

**Pull request merged**